lib/spack/spack/schema/mirrors.py

```diff
@@ -10,19 +10,59 @@
 """
 from typing import Any, Dict
 
+string_or_variable = {
+    "oneOf": [
+        {
+            "type": "object",
+            "required": ["variable"],
+            "additionalProperties": False,
+            "properties": {"variable": {"type": "string"}},
+        },
+        {"type": "string"},
+    ]
+}
+
+
 #: Common properties for connection specification
 connection = {
     "url": {"type": "string"},
     # todo: replace this with named keys "username" / "password" or "id" / "secret"
     "access_pair": {
-        "type": "array",
-        "items": {"type": ["string", "null"], "minItems": 2, "maxItems": 2},
+        "oneOf": [
+            {
+                "type": "array",
+                "items": {"minItems": 2, "maxItems": 2, **string_or_variable},
```

Comment:
Why `string_or_variable`, you've just introduced a dict option below. Pick one :p

Also you `null` was there for backward compatibility, don't delete it.

Reply:
The only reason is to allow a variable name and a static password (see `cmd/mirror.py`). That case doesn't really make sense so I think dropping it is probably fine, I was getting a bit overly pedantic.

I will make that case an error and drop the string-or-variable.